command/agentproxyshared/cache/lease_cache.go

```diff
@@ -448,12 +536,72 @@ func (c *LeaseCache) Send(ctx context.Context, req *SendRequest) (*SendResponse,
 		return nil, err
 	}
 
-	// Start renewing the secret in the response
-	go c.startRenewing(renewCtx, index, req, secret)
+	if index.Type != cacheboltdb.StaticSecretType {
+		// Start renewing the secret in the response
+		go c.startRenewing(renewCtx, index, req, secret)
+	}
 
 	return resp, nil
 }
 
+func (c *LeaseCache) cacheStaticSecret(ctx context.Context, req *SendRequest, resp *SendResponse, index *cachememdb.Index) error {
+	// If a cached version of this secret exists, we now have access, so
+	// we don't need to re-cache, just update index.Tokens
+	indexFromCache, err := c.db.Get(cachememdb.IndexNameID, index.ID)
+	if err != nil {
+		return err
+	}
+
+	// The index already exists, so all we need to do is add our token
+	// to the index's allowed token list, then re-store it
+	if indexFromCache != nil {
+		indexFromCache.Tokens = append(indexFromCache.Tokens, req.Token)
+
+		return c.storeStaticSecretIndex(ctx, req, indexFromCache)
+	}
+
+	// Serialize the response to store it in the cached index
+	var respBytes bytes.Buffer
+	err = resp.Response.Write(&respBytes)
+	if err != nil {
+		c.logger.Error("failed to serialize response", "error", err)
+		return err
+	}
+
+	// Reset the response body for upper layers to read
+	if resp.Response.Body != nil {
+		resp.Response.Body.Close()
+	}
+	resp.Response.Body = io.NopCloser(bytes.NewReader(resp.ResponseBody))
+
+	// Set the index's Response
+	index.Response = respBytes.Bytes()
+
+	// Set the index's tokens
+	index.Tokens = []string{req.Token}
+
+	// Set the index type
+	index.Type = cacheboltdb.StaticSecretType
+
+	return c.storeStaticSecretIndex(ctx, req, index)
+}
+
+func (c *LeaseCache) storeStaticSecretIndex(ctx context.Context, req *SendRequest, index *cachememdb.Index) error {
+	// Store the index in the cache
+	c.logger.Debug("storing response into the cache", "method", req.Request.Method, "path", req.Request.URL.Path)
+	err := c.Set(ctx, index)
+	if err != nil {
+		c.logger.Error("failed to cache the proxied response", "error", err)
+		return err
+	}
+
+	// TODO: We need to also update the cache for the token's permission capabilities.
+	// TODO: for this we'll need: req.Token, req.URL.Path
```

Comment:
Just in case it wasn't clear, this is also coming in the next PR, to better separate the kinds of caching and to make it easier to review. I thought I'd leave this here so it was obvious what was missing!